invoices/invoice_expiry_watcher.go

```diff
@@ -0,0 +1,178 @@
+// will add all active (not yet settled and not yet cancelled) invoices to its
+// watcing queue.
+// When a new invoice is added to the InvoiceRegistry, InvoiceExpiryWatcher
+// will add that to the watching queue as well.
+// If any of the watched invoices expire, they'll be removed from the watching
+// queue and will be cancelled through InvoiceRegistry.CancelInvoice.
+type InvoiceExpiryWatcher struct {
+	started sync.Once
+	stopped sync.Once
+	wg      sync.WaitGroup
+
+	// clock is the clock implementation that InvoiceExpiryWatcher uses.
+	// It is useful for testing.
+	clock clock.Clock
+
+	// quit signals InvoiceExpiryWatcher to stop.
+	quit chan interface{}
+
+	// cancelInvoice is a template method that cancels an expired invoice.
+	cancelInvoice func(lntypes.Hash) error
+
+	// expiryQueue holds invoiceExpiry items and is used to find the next
+	// invoice to expire.
+	expiryQueue queue.PriorityQueue
+
+	// newInvoices channel is used to wake up the main loop when a new invoices
+	// is added.
+	newInvoices chan *invoiceExpiry
+}
+
+// NewInvoiceExpiryWatcher creates a new InvoiceExpiryWatcher instance.
+func NewInvoiceExpiryWatcher(clock clock.Clock) *InvoiceExpiryWatcher {
+	return &InvoiceExpiryWatcher{
+		clock:       clock,
+		quit:        make(chan interface{}),
+		newInvoices: make(chan *invoiceExpiry),
+	}
+}
+
+// PrefetchInvoices fetches all active invoices and their corresponding payment
+// hashes from ChannelDB and adds them to the watcher. This is useful to
+// prepopulate the watcher with past, but active invoices upon start.
+func (ew *InvoiceExpiryWatcher) PrefetchInvoices(cdb *channeldb.DB) {
+	pendingOnly := true
+	invoices, hashes, err := cdb.FetchAllInvoicesWithPaymentHash(pendingOnly)
+	if err != nil {
+		log.Errorf("Error fetching invoices from the database: %v", err)
+	} else {
+		for k := 0; k < len(invoices); k++ {
+			ew.AddInvoice(hashes[k], &invoices[k])
+		}
+	}
+}
+
+// Start starts the the subscription handler and the main loop. Start() can
+// only be called once.
+func (ew *InvoiceExpiryWatcher) Start(cancelInvoice func(lntypes.Hash) error) {
+	ew.started.Do(func() {
+		ew.cancelInvoice = cancelInvoice
+		ew.wg.Add(1)
+		go ew.mainLoop()
+	})
+}
+
+// Stop cancels the invoice subscription and stops the expiry handler loop.
+func (ew *InvoiceExpiryWatcher) Stop() {
+	ew.stopped.Do(func() {
+		// Signal subscriptionHandler to quit and wait for it to return.
+		close(ew.quit)
+		ew.wg.Wait()
+	})
+}
+
+// AddInvoice adds a new invoice to the InvoiceExpiryWatcher. This won't check
+// if the invoice is already added and will only add invoices with ContractOpen
+// state.
+func (ew *InvoiceExpiryWatcher) AddInvoice(
+	paymentHash lntypes.Hash, invoice *channeldb.Invoice) {
+
+	if invoice.Terms.State != channeldb.ContractOpen {
+		log.Debugf("Invoice not added to expiry watcher: %v", invoice)
+		return
+	}
+
+	expiry := invoice.CreationDate.Add(
+		zpay32.DefaultInvoiceExpiry(invoice.Expiry))
```

Comment:
Name `DefaultInvoiceExpiry` doesn't fully cover what the function does

Reply:
PTAL